src/library/scala/concurrent/package.scala

```diff
@@ -170,7 +173,7 @@ package concurrent {
     @throws(classOf[TimeoutException])
     @throws(classOf[InterruptedException])
     final def ready[T](awaitable: Awaitable[T], atMost: Duration): awaitable.type = awaitable match {
-      case f: Future[T] if f.isCompleted => awaitable.ready(atMost)(AwaitPermission)
+      case f: Future[T] if f.isCompleted => awaitable
```

Comment:
This changes semantics. Before:

```scala
scala> import scala.concurrent._; import duration._; import ExecutionContext.Implicits.global

scala> Await.ready(Future.successful(1), duration.Duration.Undefined)
java.lang.IllegalArgumentException: Cannot wait for Undefined duration of time
```

After:

```scala
scala> Await.ready(Future.successful(1), duration.Duration.Undefined)
val res0: scala.concurrent.Future[Int] = Future(Success(1))
```

Same concern for `Await.result` below.

The difference is only for `atMost eq Undefined`, I'm fine to include that test here.

Reply:
Yes, the old one will always check the duration first, the new one follows Java 's CompletableFuture
in Java:
```
    public T get(long timeout, TimeUnit unit)
        throws InterruptedException, ExecutionException, TimeoutException {
        long nanos = unit.toNanos(timeout);
        Object r;
        if ((r = result) == null)
            r = timedGet(nanos);
        return (T) reportGet(r);
    }
```
When the value is completed, return the value.

Reply:
@pjfanning @raboof cc

Reply:
Or also this would be fine with me

```scala
      case f: Future[T] if f.isCompleted =>
        if (atMost ne Duration.Undefined) awaitable
        else throw new IllegalArgumentException("Cannot wait for Undefined duration of time")
```